src/levanter/trainer.py

```diff
@@ -636,19 +642,58 @@ def initialize(self):
 
     @cached_property
     def device_mesh(self) -> Mesh:
-        devices = jax.devices()
-        devices = np.array(devices).reshape(self.data_axis_size, self.model_axis_size)
-        return Mesh(devices, (ResourceAxis.DATA, ResourceAxis.MODEL))
+        is_multislice = hasattr(jax.devices()[0], "slice_index")
+        if is_multislice:
+            devices = mesh_utils.create_hybrid_device_mesh(
+                (self.replica_ici_axis_size, self.data_ici_axis_size, self.model_axis_size),
+                (self.replica_dcn_axis_size, self.data_dcn_axis_size, 1),
+                allow_split_physical_axes=True,
+            )
+        else:
+            devices = mesh_utils.create_device_mesh(
+                (self.replica_ici_axis_size, self.data_ici_axis_size, self.model_axis_size),
+                allow_split_physical_axes=True,
+            )
+
+        return Mesh(devices, (ResourceAxis.REPLICA, ResourceAxis.DATA, ResourceAxis.MODEL))
 
     @property
     def eval_batch_size(self):
         return self.per_device_eval_parallelism * self.data_axis_size
 
+    @property
+    def num_slices(self):
+        """number of nodes"""
+        return getattr(jax.devices()[-1], "slice_index", 0) + 1
```

Comment:
nit: I don't think it's wise to assume that these are sorted. could we max and use cached_property